### ISSUE TITLE
Classification with C++ TorchScript

### ISSUE THREAD

**Issue by joewong00** [ultralytics/yolov5#12375]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,

I'm trying to deploy a classification model that I trained using yolov5. I was able to do inferencing using the normal python way 
```
python classify/predict.py --weights best.pt --source img.jpg
```
with high accuracy and performance. However, as I'm trying to do the same in C++ using the libTorch library, I'm having much worse performance and inference time. I first exported the trained model to a torchscript model as follow:
```
python export.py --weights best.pt --include torchscript --device 0
```

After obtaining the torchscript file, I then tried to run the inference in C++ runtime with the code below:
```
torch::jit::script::Module module = torch::jit::load("best.torchscript");

cv::Mat input_img;
cv::resize(input, input_img, cv::Size(640, 384));
cv::cvtColor(input_img, input_img, cv::COLOR_BGR2RGB);
torch::Tensor imgTensor = torch::from_blob(input_img.data, {input_img.rows, input_img.cols,3},torch::kByte);
imgTensor = imgTensor.permute({2,0,1});
imgTensor = imgTensor.toType(torch::kFloat);
imgTensor = imgTensor.div(255);
imgTensor = imgTensor.unsqueeze(0).to(device);

auto start_time = std::chrono::high_resolution_clock::now();
torch::Tensor output = module.forward({imgTensor}).toTensor();

auto end_time = std::chrono::high_resolution_clock::now();
auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

std::cout << "Inference time: " << duration.count() << " milliseconds" << std::endl;
```

The code compiled and run with no issue. However I realised the inference time is super long, ironically it became worse when I run it using CUDA (about 5000+ ms) as compared to CPU (about 500+ ms). The reason I run it in C++ is to have an optimized and real time inferencing. I have no idea which part did I do wrongly and I couldn't find any resources about this online, so any help or guidance would be much appreciated!

**Comment by glenn-jocher** [comment]
@joewong00 hi there,

Thanks for reaching out! It looks like you're experiencing slower inference times when running your YOLOv5 model in C++ using libTorch compared to the Python implementation.  

It's possible that the performance difference could be due to differences in how PyTorch and C++ handle certain operations, as well as potential differences in optimization levels between the Python and C++ runtimes. I recommend checking the memory and performance profiling of your C++ code to pinpoint any possible bottlenecks. Additionally, you may want to verify that your C++ environment and CUDA are properly set up for optimal performance.

Lastly, you might find more specific advice within the PyTorch community where experts can offer insights on optimizing performance in C++ deployments.

I hope this provides some guidance! Let me know if you have any further questions or run into any issues. Good luck with your deployment!

**Comment by joewong00** [comment]
Hi @glenn-jocher ,

Thank you for your reply. I somewhat found out the issue. According to https://github.com/pytorch/pytorch/issues/47737, the classification is only slow when I'm trying to only inference an image for the first time. For some reasons, to use the GPU for the first time it took a longer time to initialize. After that it is able to detect at a very fast rate (~10ms).

Thank you.

**Comment by glenn-jocher** [comment]
You're welcome, @joewong00! I'm glad to hear you've identified the issue. It's great to know that the inference speed improved after the initial GPU initialization. If you have any more questions or need further assistance, don't hesitate to reach out. Good luck with your continued work on YOLOv5 deployment!
